### DISCUSSION TITLE
如何去掉编译前的所有flags检查呢？

### DISCUSSION THREAD

**Discussion by DanielKui** [xmake-io/xmake#3954] [Q&A] [answered]
我使用了set_policy("check.auto_ignore_flags", false) 
的确减少了一些，但是还留存了一些检查如下：

![image](https://github.com/xmake-io/xmake/assets/578118/f375402f-5d71-4555-b481-888e7fd6af4e)

如何去掉所有的flags checking呢？

**Comment by waruqi** [comment]
剩下的这些是跨编译器必须的，内部的 flags 处理。。不能去。你能去的只能是 用户自己 add_cxflags 设置的 flags 检测。。

**Comment by waruqi** [comment]
就比如支持 c++modules 特性。不同编译器 gcc clang msvc，以及不同版本，需要的 flags 都是不同的，内部需要自动检测后，是适配各种编译器支持的力度。。

其他特性也是。。没法去掉

**Comment by DanielKui** [comment]
非常感谢，我有一个sh脚本在根目录下，如下，编译每一个子目录下的项目，但每执行一次xmake都要有这样的检查，有没有办法做到只检查一次呢？

xmake -j$handleCpuNum $verbose --file=./Tools/LogManager/LogManager.lua
xmake -j$handleCpuNum $verbose --file=./HXApi/HXQuoteDemoAsio.lua
xmake -j$handleCpuNum $verbose --file=./HXApi/HXTradeDemoAsio.lua
xmake -j$handleCpuNum $verbose --file=./HXApi/HXQuoteDemo.lua
xmake -j$handleCpuNum $verbose --file=./HXApi/HXTradeDemo.lua
xmake -j$handleCpuNum $verbose --file=./HXApi/HXRiskDemo.lua

**Reply by waruqi** [reply]
你这样拆成独立进程，肯定会多次检测，避免不了。。而且这也是不推荐的。干嘛不用 includes + xmake build targetname 呢？或者用 target group 。。直接 `xmake build -g groupname` 。。只要保证在一个进程，就能减少检测次数

**Reply by DanielKui** [reply]
太酷了，新思路

-j$handleCpuNum $verbose   这几个参数是想在命令行中用户指定的，如何传递给这个a.lua呢？

a.lua

includes("./Tools/LogManager/LogManager.lua")
includes("./HXApi/HXQuoteDemoAsio.lua")
includes("./HXApi/HXTradeDemoAsio.lua")
includes("./HXApi/HXQuoteDemo.lua")
includes("./HXApi/HXTradeDemo.lua")
includes("./HXApi/HXRiskDemo.lua")
includes("./QuoteSvr/QuoteSvr_EQ_Asio.lua")
includes("./QuoteSvr/QuoteSvr_ST_Asio.lua")
includes("./QuoteSvr/QuoteSvr_XTP_Asio.lua")
includes("./QuoteSvr/QuoteSvr_Insight_Asio.lua")
includes("./ETFSvr/ETFSvr_DD.lua")
includes("./ETFSvr/ETFSvr_XTP.lua")
includes("./ETFSvr/ETFSvr_COS.lua")
includes("./RiskCtrlSvr/RiskCtrlSvr.lua")
includes("./TradeSvr/TradeSvr_COS.lua")
includes("./TradeSvr/TradeSvr_DD.lua")
includes("./TradeSvr/TradeSvr_XTP.lua")
includes("./TradeSvr/TradeSvr_JG.lua")


xmake build -j16 -verbose LogManager
xmake build -j16 -verbose XXXXX

**Reply by waruqi** [reply]
为什么要具体传给某个 a.lua 呢。。lua 的目的就是维护构建 target。。你 build target 时候，指定 -jX 不就好了么

**Reply by DanielKui** [reply]
![image](https://github.com/xmake-io/xmake/assets/578118/7f2ef7c6-17c1-4e15-8aee-4a28eec65446)

我在sh脚本里，如何获取target name 呢？好像includes是lua的语法，我有点乱了

**Reply by waruqi** [reply]
不是啊。。你为啥不项目根目录搞个 xmake.lua 。。用 includes 把这些 lua 全组织进来呢。。还去写 shell 干嘛。。用 xmake 的目的就是 0 shell

**Reply by waruqi** [reply]
仔细看下文档 https://xmake.io/#/zh-cn/guide/syntax_description?id=%e5%a4%9a%e7%ba%a7%e9%85%8d%e7%bd%ae
https://xmake.io/#/zh-cn/guide/syntax_description?id=%e5%a4%9a%e7%ba%a7%e9%85%8d%e7%bd%ae

**Comment by DanielKui** [comment]
> 不是啊。。你为啥不项目根目录搞个 xmake.lua 。。用 includes 把这些 lua 全组织进来呢。。还去写 shell 干嘛。。用 xmake 的目的就是 0 shell
哦哦，那这个根目录里的xmake.lua 如何获取-jX呢？我是想获取用户指定核数，或者满核编译。 以及--verbose等选项

**Reply by waruqi** [reply]
> 哦哦，那这个根目录里的xmake.lua 如何获取-jX呢？我是想获取用户指定核数，或者满核编译。 以及--verbose等选项

干嘛要获取 -jX 呢？ 你默认 xmake 编译，啥参数不带，就是满核编译。。不要自己搞复杂。。

**Reply by DanielKui** [reply]
![image](https://github.com/xmake-io/xmake/assets/578118/46e63d0b-f14c-446c-ad57-580a48da2370)

恕我不太懂啊，这是我的xmake.lua，xmake的时候，出现这个错误了。。。。

**Reply by waruqi** [reply]
。。。 xmake build 是命令，你放 xmake.lua 干嘛。。建议你先仔细看下文档。。

**Comment by DanielKui** [comment]
我是想搞一个xmake.lua  includes所有的子工程lua， xmake build targetname来编译的话，这个xmake build target name是写在哪里呢？我之前是写在sh里，如果不用sh的话，该怎么写呢？

**Reply by waruqi** [reply]
你只是命令行敲 xmake build xx 不就好了么。。

**Reply by waruqi** [reply]
不想挨个敲，就走分组，看文档 set_group

**Comment by DanielKui** [comment]
xmake.lua里includes进来后
在命令行中直接 xmake build -all 就可以，是这个意思吗？这个时候xmake已经识别了xmake.lua里的所有includes进来的target了

**Reply by waruqi** [reply]
看文档啊，都有详细说明

**Comment by DanielKui** [comment]
再请教一个问题哈，我在根目录下有一个xmake.lua里面includes各个子目录的sub.lua，但是子目录里有before_link  os.cp(../" 类似的拷贝操作，然后运行xmake build --file=xmake.lua后，会导致目录找不到的问题，这个该怎么处理呢？xmake.lua里全部都是includes ("HXAPI/sub.lua")类似的包含语句，好像也没法改变当前目录

这是我的xmake.lua
set_policy("check.auto_ignore_flags", false)
includes("./Tools/LogManager/LogManager.lua")
includes("./HXApi/HXQuoteDemoAsio.lua")
includes("./HXApi/HXTradeDemoAsio.lua")
includes("./HXApi/HXQuoteDemo.lua")
includes("./HXApi/HXTradeDemo.lua")
includes("./HXApi/HXRiskDemo.lua")
includes("./QuoteSvr/QuoteSvr_EQ_Asio.lua")
includes("./QuoteSvr/QuoteSvr_ST_Asio.lua")
includes("./QuoteSvr/QuoteSvr_XTP_Asio.lua")
includes("./QuoteSvr/QuoteSvr_Insight_Asio.lua")
includes("./ETFSvr/ETFSvr_DD.lua")
includes("./ETFSvr/ETFSvr_XTP.lua")
includes("./ETFSvr/ETFSvr_COS.lua")
includes("./RiskCtrlSvr/RiskCtrlSvr.lua")
includes("./TradeSvr/TradeSvr_COS.lua")
includes("./TradeSvr/TradeSvr_DD.lua")
includes("./TradeSvr/TradeSvr_XTP.lua")
includes("./TradeSvr/TradeSvr_JG.lua")


![image](https://github.com/xmake-io/xmake/assets/578118/82e70686-f29a-45ba-90b4-d9bd95b5fbff)

**Comment by waruqi** [comment]
> 然后运行xmake build --file=xmake.lua后

你为啥非要指定 `--file=xmake.lua` 呢，不嫌麻烦么。。直接 `xmake` 不行么

> 会导致目录找不到的问题，这个该怎么处理呢？xmake.lua里全部都是includes ("HXAPI/sub.lua")类似的包含语句，好像也没法改变当前目录

干嘛要改变目录。。脚本域自己 `path.join(os.projectdir(), "xxx")` 按项目根目录处理不就好了。

**Comment by DanielKui** [comment]
kb@devmachine:~/cpp1/trunk$ xmake
[100%]: build ok, spent 0.006s
warning: includes("./Tools/LogManager/LogManager.lua") cannot find any files!
warning: add -v for getting more warnings ..


我直接xmake 后，出现这个错误。
但--file=xmake.lua就没这个错误，奇怪




kb@devmachine:~/cpp3/trunk$ xmake build -a --verbose
[100%]: build ok, spent 0.006s
warning: includes("./Tools/LogManager/LogManager.lua") cannot find any files!
warning: includes("./HXApi/HXQuoteDemoAsio.lua") cannot find any files!
warning: includes("./HXApi/HXTradeDemoAsio.lua") cannot find any files!
warning: includes("./HXApi/HXQuoteDemo.lua") cannot find any files!
warning: includes("./HXApi/HXTradeDemo.lua") cannot find any files!
warning: includes("./HXApi/HXRiskDemo.lua") cannot find any files!
warning: includes("./QuoteSvr/QuoteSvr_EQ_Asio.lua") cannot find any files!
warning: includes("./QuoteSvr/QuoteSvr_ST_Asio.lua") cannot find any files!
warning: includes("./QuoteSvr/QuoteSvr_XTP_Asio.lua") cannot find any files!
warning: includes("./QuoteSvr/QuoteSvr_Insight_Asio.lua") cannot find any files!
warning: includes("./ETFSvr/ETFSvr_DD.lua") cannot find any files!
warning: includes("./ETFSvr/ETFSvr_XTP.lua") cannot find any files!
warning: includes("./ETFSvr/ETFSvr_COS.lua") cannot find any files!
warning: includes("./RiskCtrlSvr/RiskCtrlSvr.lua") cannot find any files!
warning: includes("./TradeSvr/TradeSvr_COS.lua") cannot find any files!
warning: includes("./TradeSvr/TradeSvr_DD.lua") cannot find any files!
warning: includes("./TradeSvr/TradeSvr_XTP.lua") cannot find any files!
warning: includes("./TradeSvr/TradeSvr_JG.lua") cannot find any files!

**Reply by waruqi** [reply]
路径不匹配，自己对下

**Comment by DanielKui** [comment]
path.join(os.projectdir(), "xxx")

如果这么做，单独到子目录里编译子项目的话，子项目的路径又不对了。

根目录里统一编译和 单独到子目录里编译，都可以找到正确路径，这个该怎么处理呢？
实在是穷尽了，没招了我，投降，请教大佬


![image](https://github.com/xmake-io/xmake/assets/578118/8b9d9af2-a4b4-42ac-be6c-1ce5490b7c37)

**Comment by waruqi** [comment]
> 根目录里统一编译和 单独到子目录里编译，都可以找到正确路径，这个该怎么处理呢？
实在是穷尽了，没招了我，投降，请教大佬

我也对你没招了，各种瞎搞。。建议先多看看文档。。

os.projectdir(), os.targetdir() 都是绝对路径，能 join 到一起？

ln 有 os.ln ，为啥要执行 os.run ln 。。里面路径全不对。。

**Reply by DanielKui** [reply]
哦哦，那个ln 是我乱写的，但问题是根目录和子目录的os.projectdir()是不同的，如果我单独到子目录里编译的话，运行os.cp的时候 projectdir()又变了。
os.projectdir() 在根目录下运行的时候是/home/kb/trunk, 在子目录下运行的时候是/home/kb/trunk/HXApi了

正确的路径是/home/kb/trunk/Tools/LogManager/bin/libLogManager_x64.so

如果在/home/kb/trunk/HXApi目录里运行xmake --file=子目录项目的lua 的话，则os.projectdir() = /home/kb/trunk/HXApi， 那么os.cp拷贝的时候就去找/home/kb/trunk/HXApi/Tools/LogManager/bin/libLogManager_x64.so(实际上的路径应该是/home/kb/trunk/Tools/LogManager/bin/libLogManager_x64.so )

os.cp(path.join(os.projectdir(), "/Tools/LogManager/bin/libLogManager_x64.so"), target:targetdir())

**Reply by waruqi** [reply]
> projectdir()又变了。

这个是不会变的，xmake.lua 在项目根目录，那么目录永远是这个。。除非你自己 --file=xxx.lua 切了。。所以不让你用 --file改项目跟目录。。

用 includes 。。

**Reply by waruqi** [reply]
> 根目录和子目录的os.projectdir()是不同的

走 includes. os.projectdir 永远只有一个。。

**Reply by DanielKui** [reply]
对问题就在这里，子目录下很多子项目的话，是编辑子目录下的xmake.lua 只includes()自己想要编译的lua吗?是这个意思不

**Reply by waruqi** [reply, marked as answer]
你怎么还没理解呢。。我记得文档里说的很详细的。。先自己仔细看看哈。。

你不管有几个子目录，几个子项目。反正就一个根目录。一个根 xmake.lua 里面 includes 把所有子目录的 lua 全部 include 进来。。

就只有一个大的单一工程，里面有多少 target 都无所谓。。自己 xmake build targetxxx 自己想编译哪些 target 就编译哪些。。所有命令都只在跟目录下执行。。也不要切目录。。懂了么。。。。反正就是搞成一个工程。。没那么多事

**Reply by DanielKui** [reply]
跪谢大佬，来青岛我请你喝啤酒

**Comment by DanielKui** [comment]
![image](https://github.com/xmake-io/xmake/assets/578118/83a93b39-ec25-47aa-8f16-bb4ce45a67bf)

请教下，这里我用xmake build targetname单独编译一个target，奇怪为何找不到.o呢？
 ll build/.objs/HXQuoteApi/linux/x86_64/release/common/HXPtclPacket.cpp.o
也存在
